File: 2 SEGUNDO/1 Cuatrimestre/Programación y Diseño Orientado a Objetos - PDOO/Practicas/rubyNapakalaki/specific_bad_consequence.rb
class SpecificBadConsequence < Bad_consequence

  attr_accessor :specificVisibleTreasures , :specificHiddenTreasures


  def initialize(text,levels,tVisible,tHidden)
        super(text, levels)
        @specificHiddenTreasures = tHidden
        @specificVisibleTreasures = tVisible
  end

  def getVisibleTreasures
       @specificVisibleTreasures
  end

  def getHiddenTreasures
       @specificHiddenTreasures
  end

  def to_s
    "#{@text} Pierdes: #{@levels} niveles #{@specificVisibleTreasures} tesoros visibles y #{@specificHiddenTreasures} tesoros ocultos"
  end

  def is_dead
       var = false
       var
  end

  def is_empty
       var = false
       if @text == "" and @levels == 0 and @specificVisibleTreasures.empty? and @specificHiddenTreasures.empty?
            var = true
         end

       var

  end

  def setnVisibleTreasures(n)
      raise NotImplementedError.new
  end

  def setnHiddenTreasures(n)
      raise NotImplementedError.new
  end

  def getnVisibleTreasures
      return 0
  end

  def getnHiddenTreasures
      return 0
  end

  def substract_visible_treasure(t)
      @specificVisibleTreasure.delete(t.type)
  end

  def substract_hidden_treasure(t)
      @specificHiddenTreasure.delete(t.type)
  end

def adjust_to_fit_treasure_list(v, h)
     visible = Array.new
     hidden = Array.new

        v.each do |t|

            if @specificVisibleTreasures.include?(t.type) == true
                visible << t.type
            end
        end


        h.each do |t|

            if @specificHiddenTreasures.include?(t.type) == true
                hidden << t.type
            end
        end

        bad_cons = SpecificBadConsequence.new('', @levels, visible, hidden)
        bad_cons
  end


end

File: 2 SEGUNDO/1 Cuatrimestre/Programación y Diseño Orientado a Objetos - PDOO/Practicas/javaNapakalaki/NapakalakiGame/SpecificBadConsequence.java
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author manjaro
*/

public class SpecificBadConsequence extends BadConsequence {
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();

    public SpecificBadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden) {
        super(text, levels);
        this.specificHiddenTreasures = tHidden;
        this.specificVisibleTreasures = tVisible;
    }

    public ArrayList<TreasureKind> getVisibleTreasures(){

        return specificVisibleTreasures;
    }

    public ArrayList<TreasureKind> getHiddenTreasures(){

        return specificHiddenTreasures;
    }

    private String ArrayToString(ArrayList<TreasureKind> datos){
        String uno="" ;

        for (TreasureKind t: datos){
            uno += t.toString() + "\t";
        }
        return uno;

    }

     @Override
     public boolean isDeath(){
       return false;
     }

     @Override
     public void setnVisibleTreasures(int n){}

     @Override
     public void setnHiddenTreasures(int n){}

     @Override
     public int getnVisibleTreasures(){
         return 0;
     }

     @Override
     public int getnHiddenTreasures(){
         return 0;
     }

     @Override
     public void substractVisibleTreasure(Treasure t){

        for(int i =0; i< specificVisibleTreasures.size(); i++){
             if(t.getType() == specificVisibleTreasures.get(i)){
                 specificVisibleTreasures.remove(i);
             }
        }
     }

     @Override
     public void substractHiddenTreasure(Treasure t){

         for(int i =0; i< specificHiddenTreasures.size(); i++){
             if(t.getType() == specificHiddenTreasures.get(i)){
                 specificHiddenTreasures.remove(i);

             }
        }
     }

     @Override
     public  BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v , ArrayList<Treasure> h){

         BadConsequence uno;

          ArrayList<TreasureKind> visibles = new ArrayList();
          ArrayList<TreasureKind> ocultos = new ArrayList();

           for(Treasure iterador:v){
               if(this.specificVisibleTreasures.contains(iterador.getType())){
                   visibles.add(iterador.getType());
               }
           }

          for(Treasure iterador:h){
               if(this.specificHiddenTreasures.contains(iterador.getType())){
                   ocultos.add(iterador.getType());
               }
           }
           uno =  new SpecificBadConsequence("",super.getLevels(),visibles,ocultos);
          return uno;
        }

     @Override
       public boolean isEmpty(){
           boolean uno = false;
           if(super.getText() == "" && super.getLevels() == 0 && specificHiddenTreasures.isEmpty() && specificVisibleTreasures.isEmpty() ){
              uno = true;
           }
          return uno;
       }

     @Override
     public String toString(){

        String uno = super.getText() + "\n Pierdes " + super.getLevels() + " niveles "
                +  "\nTesoros visibles: " + ArrayToString(specificVisibleTreasures)
                + "\nTesoros ocultos: " + ArrayToString(specificHiddenTreasures) + "\n";
          return uno;
      }

     public String toStringPendingVisible(){
         String uno = ArrayToString(specificVisibleTreasures);
         return uno;
     }

     public String toStringPendingOculto(){
         String uno = ArrayToString(specificHiddenTreasures);
         return uno;
     }

}
